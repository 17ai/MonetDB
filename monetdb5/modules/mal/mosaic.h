/*
 * The contents of this file are subject to the MonetDB Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.monetdb.org/Legal/MonetDBLicense
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is the MonetDB Database System.
 *
 * The Initial Developer of the Original Code is CWI.
 * Portions created by CWI are Copyright (C) 1997-July 2008 CWI.
 * Copyright August 2008-2014 MonetDB B.V.
 * All Rights Reserved.
 */

/* (c) M Kersten
 * The multi-purpose compressor infrastructure
 */
#ifndef _MOSLIST_H
#define _MOSLIST_H
#include "mal_interpreter.h"
#include "mal_exception.h"
#include "mal_function.h"

#include "mtime.h"
#include "math.h"
#include "opt_prelude.h"
#include "algebra.h"

//#define _DEBUG_MOSAIC_

#define MOSAIC_VERSION 20140808

/* do not invest in compressing BATs smaller than this */
#define MIN_INPUT_COUNT 1

/* The compressor kinds currently hardwired */
#define MOSAIC_METHODS	6
#define MOSAIC_NONE     0		// no compression at all
#define MOSAIC_RLE      1		// use run-length encoding
#define MOSAIC_DICT     2		// local dictionary encoding
#define MOSAIC_DELTA	3		// use delta encoding
#define MOSAIC_LINEAR 	4		// use an encoding for a linear sequence
#define MOSAIC_ZONE		5		// adaptive zone map over non-compressed data
#define MOSAIC_EOL		6		// marker for the last block

//Compression should have a significant reduction to apply.
#define COMPRESS_THRESHOLD 50   //percent

/*
 * The header is reserved for meta information, e.g. oid indices.
 * The block header encodes the information needed for the chunk decompressor
 */
#define MOSAICINDEX 4  //> 2 elements
typedef struct MOSAICHEADER{
	int version;
	int top;
	oid index[MOSAICINDEX];
	BUN offset[MOSAICINDEX];
} * MosaicHdr;

// bit stuffed header block
typedef lng *MosaicBlk;
#define MOStag(Blk) (*(Blk)>>56)
#define MOSsetTag(Tag)  ((lng) (Tag) <<56)
#define MOScnt(Blk) (BUN)(*(Blk) & 03777777777777777)
#define MOSinc(Blk,I) *(Blk)= *(Blk)+I

#define MOSnone (((lng)MOSAIC_NONE) <<56)
#define MOSrle (((lng)MOSAIC_RLE) <<56)
#define MOSdict (((lng)MOSAIC_DICT) <<56)
#define MOSlinear (((lng)MOSAIC_LINEAR) <<56)
#define MOSdelta (((lng)MOSAIC_DELTA) <<56)
#define MOSzone (((lng)MOSAIC_ZONE) <<56)
#define MOSeol (((lng)MOSAIC_EOL) <<56)

#define wordaligned(SZ) \
	 ((SZ) +  ((SZ) % sizeof(int)? sizeof(int) - ((SZ)%sizeof(int)) : 0))

#define MosaicHdrSize  wordaligned(sizeof(struct MOSAICHEADER))
#define MosaicBlkSize  wordaligned(sizeof(lng))


typedef struct MOSTASK{
	int type;		// one of the permissible types
	MosaicHdr hdr;	// start of the destination heap
	MosaicBlk blk;	// current block header
	char *dst;		// write pointer into current compressed blocks

	BUN	elm;		// elements left to compress
	char *src;		// read pointer into source

	lng  xsize,size;// original and compressed size
	lng timer;		// compression time
	void *min, *max;// space for zones

	oid *lb, *rb;	// Collected oids from operations
	oid *cl;		// candidate admin
	lng	n;			// element count in candidate list

	BAT *lbat, *rbat; // for the joins, where we dont know their size upfront

	// collect compression statistics for the particular task
	lng blks[MOSAIC_METHODS];	
	lng elms[MOSAIC_METHODS];	
} *MOStask;

/* we keep a condensed OID index anchored to the compressed blocks */

typedef struct MOSINDEX{
	lng offset;		// header location within compressed heap
	lng nullcnt;	// number of nulls encountered
	ValRecord low,hgh; // zone value markers for fix-length types
} *mosaicindex;

/* Run through a column to produce a compressed version */

#ifdef _MSC_VER
#define nextafter   _nextafter
float nextafterf(float x, float y);
#endif

#define PREVVALUEbit(x) ((x) - 1)
#define PREVVALUEbte(x) ((x) - 1)
#define PREVVALUEsht(x) ((x) - 1)
#define PREVVALUEint(x) ((x) - 1)
#define PREVVALUElng(x) ((x) - 1)
#define PREVVALUEoid(x) ((x) - 1)
#define PREVVALUEflt(x) nextafterf((x), -GDK_flt_max)
#define PREVVALUEdbl(x) nextafter((x), -GDK_dbl_max)

#define NEXTVALUEbit(x) ((x) + 1)
#define NEXTVALUEbte(x) ((x) + 1)
#define NEXTVALUEsht(x) ((x) + 1)
#define NEXTVALUEint(x) ((x) + 1)
#define NEXTVALUElng(x) ((x) + 1)
#define NEXTVALUEoid(x) ((x) + 1)
#define NEXTVALUEflt(x) nextafterf((x), GDK_flt_max)
#define NEXTVALUEdbl(x) nextafter((x), GDK_dbl_max)

// skip until you hit a candidate
#define MOSskipit()\
if ( task->cl && task->n){\
	while( *task->cl < (oid) first)\
		{task->cl++; task->n--;}\
	if ( *task->cl > (oid) first  || task->n ==0)\
		continue;\
	if ( *task->cl == (oid) first ){\
		task->cl++; task->n--;\
	}\
} else if (task->cl) continue;

#ifdef WIN32
#if !defined(LIBMAL) && !defined(LIBATOMS) && !defined(LIBKERNEL) && !defined(LIBMAL) && !defined(LIBOPTIMIZER) && !defined(LIBSCHEDULER) && !defined(LIBMONETDB5)
#define mosaic_export extern __declspec(dllimport)
#else
#define mosaic_export extern __declspec(dllexport)
#endif
#else
#define mosaic_export extern
#endif

mosaic_export str MOScompress(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mosaic_export str MOScompressInternal(Client cntxt, int *ret, int *bid, str properties);
mosaic_export str MOSdecompress(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mosaic_export str MOSdecompressInternal(Client cntxt, int *ret, int *bid);
mosaic_export str MOSanalyse(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mosaic_export str MOSsubselect(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mosaic_export str MOSthetasubselect(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mosaic_export str MOSleftfetchjoin(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mosaic_export str MOSjoin(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mosaic_export str MOSdump(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);

#endif /* _MOSLIST_H */
