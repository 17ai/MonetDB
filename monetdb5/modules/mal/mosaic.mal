module mosaic;

pattern compress(b:bat[:oid,:any_1], prop:str) :bat[:oid,:any_1]
address MOScompress
comment "Apply dynamic compression over chunks";

pattern compress(b:bat[:oid,:any_1]) :bat[:oid,:any_1]
address MOScompress
comment "Apply dynamic compression over chunks";

pattern decompress(b:bat[:oid,:any_1]) :bat[:oid,:any_1]
address MOSdecompress
comment "Apply dynamic compression over chunks";

pattern analyse()
address MOSanalyse
comment "Apply heap compression on all";

pattern dump(b:bat[:oid,:any])
address MOSdump
comment "Dump the compressed structure";

pattern subselect(b:bat[:oid,:bte], low:bte, high:bte, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:sht], low:sht, high:sht, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:int], low:int, high:int, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:lng], low:lng, high:lng, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:flt], low:flt, high:flt, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:dbl], low:dbl, high:dbl, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:timestamp], low:timestamp, high:timestamp, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect
comment "Overloaded selection routine";

pattern subselect(b:bat[:oid,:bte], cand:bat[:oid,:oid], low:bte, high:bte, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:sht], cand:bat[:oid,:oid], low:sht, high:sht, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:int], cand:bat[:oid,:oid], low:int, high:int, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:lng], cand:bat[:oid,:oid], low:lng, high:lng, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:flt], cand:bat[:oid,:oid], low:flt, high:flt, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:dbl], cand:bat[:oid,:oid], low:dbl, high:dbl, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect;

pattern subselect(b:bat[:oid,:timestamp], cand:bat[:oid,:oid], low:timestamp, high:timestamp, li:bit, hi:bit, anti:bit) :bat[:oid,:oid]
address MOSsubselect
comment "Overloaded selection routine";

pattern thetasubselect(b:bat[:oid,:bte], low:bte, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:sht], low:sht, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:int], low:int, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:lng], low:lng, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:flt], low:flt, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:dbl], low:dbl, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:timestamp], low:timestamp, oper:str) :bat[:oid,:oid]
address MOSthetasubselect
comment "Overloaded selection routine";

pattern thetasubselect(b:bat[:oid,:bte], c:bat[:oid,:oid], low:bte, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:sht], c:bat[:oid,:oid], low:sht, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:int], c:bat[:oid,:oid], low:int, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:lng], c:bat[:oid,:oid], low:lng, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:flt], c:bat[:oid,:oid], low:flt, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:dbl], c:bat[:oid,:oid], low:dbl, oper:str) :bat[:oid,:oid]
address MOSthetasubselect;

pattern thetasubselect(b:bat[:oid,:timestamp], c:bat[:oid,:oid], low:timestamp, oper:str) :bat[:oid,:oid]
address MOSthetasubselect
comment "Overloaded selection routine";

pattern leftfetchjoin(b:bat[:oid,:oid], cand:bat[:oid,:bte]) :bat[:oid,:bte]
address MOSleftfetchjoin;

pattern leftfetchjoin(b:bat[:oid,:oid], cand:bat[:oid,:sht]) :bat[:oid,:sht]
address MOSleftfetchjoin;

pattern leftfetchjoin(b:bat[:oid,:oid], cand:bat[:oid,:int]) :bat[:oid,:int]
address MOSleftfetchjoin;

pattern leftfetchjoin(b:bat[:oid,:oid], cand:bat[:oid,:lng]) :bat[:oid,:lng]
address MOSleftfetchjoin;

pattern leftfetchjoin(b:bat[:oid,:oid], cand:bat[:oid,:flt]) :bat[:oid,:flt]
address MOSleftfetchjoin;

pattern leftfetchjoin(b:bat[:oid,:oid], cand:bat[:oid,:dbl]) :bat[:oid,:dbl]
address MOSleftfetchjoin;

pattern leftfetchjoin(b:bat[:oid,:oid], cand:bat[:oid,:timestamp]) :bat[:oid,:timestamp]
address MOSleftfetchjoin
comment "Overloaded leftfetchjoin operation";

pattern join(b:bat[:oid,:any_1], gen:bat[:oid,:any_1]) (l:bat[:oid,:oid],r:bat[:oid,:oid])
address MOSjoin
comment "Overloaded join operation";
