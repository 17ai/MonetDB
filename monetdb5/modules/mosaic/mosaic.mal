module mosaic;

pattern compress(b:bat[:any_1], prop:str):bat[:any_1]
address MOScompress
comment "Apply dynamic compression over chunks";

pattern compress(b:bat[:any_1]) :bat[:any_1]
address MOScompress
comment "Apply dynamic compression producing a temporary";

pattern decompress(b:bat[:any_1]) :bat[:any_1]
address MOSdecompress
comment "Apply dynamic decompression over chunks producing a temporary version";

pattern decompressStorage(b:bat[:any_1]) :bat[:any_1]
address MOSdecompressStorage
comment "Apply dynamic decompression, changing the storage layout permanently";

pattern analyse()
address MOSanalyse
comment "Apply heap compression on all";

pattern analyse(properties:str)
address MOSanalyse
comment "Apply heap compression on all with minimum threshold";

pattern analyse(properties:str,b:int)
address MOSanalyse
comment "Apply heap compression on a single column";

pattern analyse(b:int)
address MOSanalyse
comment "Apply default heap compression on a specific column";

pattern optimize(bid:int)
address MOSoptimizer
comment "Perform a search thru the compression space using depth 1";

pattern optimize(bid:int)
address MOSoptimizer
comment "Perform a search thru the compression space of
candidate combinations";

pattern optimize(bid:int, blk:int)
address MOSoptimizer
comment "Perform a search thru the compression space of
candidate combinations";

pattern slice(bid:lng)(b0:bte,b1:bte,b2:bte,b3:bte,b4:bte,b5:bte,b6:bte,b7:bte)
address MOSslice
comment "Break column into verticalslices";

pattern dump(b:bat[:any])
address MOSdump
comment "Dump the compressed structure";

pattern select(b:bat[:any_1], low:any_1, high:any_1, li:bit, hi:bit, anti:bit) :bat[:oid]
address MOSselect
comment "Overloaded selection routine";

pattern select(b:bat[:any_1], cand:bat[:oid], low:any_1, high:any_1, li:bit, hi:bit, anti:bit) :bat[:oid]
address MOSselect
comment "Overloaded selection routine";

pattern thetaselect(b:bat[:any_1], low:any_1, oper:str) :bat[:oid]
address MOSthetaselect
comment "Overloaded selection routine";

pattern thetaselect(b:bat[:any_1], c:bat[:oid], low:any_1, oper:str) :bat[:oid]
address MOSthetaselect
comment "Overloaded selection routine";

pattern projection(b:bat[:oid], cand:bat[:any_1]) :bat[:any_1]
address MOSprojection
comment "Overloaded projection operation";

pattern join(l:bat[:any_1],r:bat[:any_1],sl:bat[:oid],sr:bat[:oid],nil_matches:bit,estimate:lng) (:bat[:oid],:bat[:oid])
address MOSjoin
comment "Overloaded join operation";

