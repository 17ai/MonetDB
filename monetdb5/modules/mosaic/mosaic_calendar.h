/*
 * The contents of this file are subject to the MonetDB Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.monetdb.org/Legal/MonetDBLicense
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is the MonetDB Database System.
 *
 * The Initial Developer of the Original Code is CWI.
 * Portions created by CWI are Copyright (C) 1997-July 2008 CWI.
 *                * Copyright August 2008-2014 MonetDB B.V.
 * All Rights Reserved.
 */

/*
 * (c)2014 author Martin Kersten
 */

#ifndef _MOSAIC_TEMPORAL_
#define _MOSAIC_TEMPORAL_

#include <mal.h>
#include "mal_interpreter.h"
#include "mal_client.h"

#define TEMPORALTHRESHOLD 4192
#define TEMPORALSIZE 256 

mal_export void MOScreatecalendar(Client cntxt, MOStask task);
mal_export void MOSdump_calendar(Client cntxt, MOStask task);
mal_export void MOSlayout_calendar(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSlayout_calendar_hdr(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSadvance_calendar(Client cntxt, MOStask task);
mal_export void MOSskip_calendar(Client cntxt, MOStask task);
mal_export flt  MOSestimate_calendar(Client cntxt, MOStask task);
mal_export void MOScompress_calendar(Client cntxt, MOStask task);
mal_export void MOSdecompress_calendar(Client cntxt, MOStask task);
mal_export str MOSselect_calendar(Client cntxt,  MOStask task, void *low, void *hgh, bit *li, bit *hi, bit *anti);
mal_export str MOSthetaselect_calendar(Client cntxt,  MOStask task, void *val, str oper);
mal_export str MOSprojection_calendar(Client cntxt,  MOStask task);
mal_export str MOSjoin_calendar(Client cntxt,  MOStask task);
#endif /* _MOSAIC_TEMPORAL_ */
