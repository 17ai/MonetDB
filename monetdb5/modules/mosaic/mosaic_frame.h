/*
 * The contents of this file are subject to the MonetDB Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.monetdb.org/Legal/MonetDBLicense
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is the MonetDB Database System.
 *
 * The Initial Developer of the Original Code is CWI.
 * Portions created by CWI are Copyright (C) 1997-July 2008 CWI.
 *                * Copyright August 2008-2014 MonetDB B.V.
 * All Rights Reserved.
 */

/*
 * (c)2014 author Martin Kersten
 */

#ifndef _MOSAIC_FRAME_
#define _MOSAIC_FRAME_

#include <mal.h>
#include "mal_interpreter.h"
#include "mal_client.h"
#include "mosaic_dictionary.h"

mal_export void MOScreateframeDictionary(Client cntxt, MOStask task);
mal_export void MOSdump_frame(Client cntxt, MOStask task);
mal_export void MOSlayout_frame_hdr(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSlayout_frame(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSadvance_frame(Client cntxt, MOStask task);
mal_export void MOSskip_frame(Client cntxt, MOStask task);
mal_export flt  MOSestimate_frame(Client cntxt, MOStask task);
mal_export void MOScompress_frame(Client cntxt, MOStask task);
mal_export void MOSdecompress_frame(Client cntxt, MOStask task);
mal_export str MOSsubselect_frame(Client cntxt,  MOStask task, void *low, void *hgh, bit *li, bit *hi, bit *anti);
mal_export str MOSthetasubselect_frame(Client cntxt,  MOStask task, void *val, str oper);
mal_export str MOSprojection_frame(Client cntxt,  MOStask task);
mal_export str MOSsubjoin_frame(Client cntxt,  MOStask task);
#endif /* _MOSAIC_FRAME_ */
