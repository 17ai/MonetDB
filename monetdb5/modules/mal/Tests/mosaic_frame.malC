b:= bat.new(:oid,:timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
bat.append(b,"2014-08-14 12:01:00":timestamp);
bat.append(b,"2014-08-14 12:02:00":timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
b:= bat.append(b,b);
base:= algebra.copy(b);

x:= mosaic.compress(b,"frame");
#mosaic.dump(x);
z:= mosaic.decompress(x);
io.print(base,z);

d:= bat.new(:oid,:date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-15":date);
d:= bat.append(d,d);

xd:= mosaic.compress(d,"frame");
#mosaic.dump(xd);
zd:= mosaic.decompress(xd);
io.print(d,zd);

e:= bat.new(:oid,:date);
bat.append(e,"1996-03-13":date);
bat.append(e,"1996-04-12":date);
bat.append(e,"1996-01-29":date);
bat.append(e,"1996-04-21":date);
bat.append(e,"1996-03-30":date);
bat.append(e,"1996-01-30":date);
bat.append(e,"1997-01-28":date);
bat.append(e,"1994-02-02":date);
bat.append(e,"1993-11-09":date);
bat.append(e,"1994-01-16":date);
bat.append(e,"1993-12-04":date);
bat.append(e,"1993-12-14":date);
bat.append(e,"1993-10-29":date);
bat.append(e,"1996-01-10":date);
bat.append(e,"1994-10-31":date);
bat.append(e,"1994-10-16":date);
bat.append(e,"1994-08-08":date);
bat.append(e,"1992-04-27":date);
bat.append(e,"1996-05-07":date);
bat.append(e,"1996-02-01":date);
bat.append(e,"1996-01-15":date);
bat.append(e,"1996-03-21":date);
bat.append(e,"1996-02-11":date);
bat.append(e,"1996-01-16":date);
bat.append(e,"1996-02-10":date);
bat.append(e,"1995-10-23":date);
bat.append(e,"1995-08-14":date);
bat.append(e,"1995-08-07":date);
bat.append(e,"1995-08-04":date);
bat.append(e,"1995-08-28":date);
bat.append(e,"1995-07-21":date);
bat.append(e,"1993-10-29":date);
bat.append(e,"1993-12-09":date);
bat.append(e,"1993-12-09":date);
bat.append(e,"1993-11-09":date);
bat.append(e,"1998-10-23":date);
bat.append(e,"1998-10-09":date);
bat.append(e,"1998-10-30":date);
bat.append(e,"1996-02-21":date);
bat.append(e,"1996-01-22":date);
bat.append(e,"1996-01-19":date);
bat.append(e,"1995-11-26":date);
bat.append(e,"1995-11-08":date);
bat.append(e,"1996-02-01":date);
bat.append(e,"1996-02-03":date);
bat.append(e,"1992-07-21":date);
bat.append(e,"1992-07-02":date);
bat.append(e,"1992-07-10":date);
bat.append(e,"1996-09-29":date);
bat.append(e,"1996-11-14":date);
bat.append(e,"1996-11-04":date);
bat.append(e,"1996-09-26":date);
bat.append(e,"1996-10-02":date);
bat.append(e,"1996-10-17":date);
bat.append(e,"1996-12-08":date);
bat.append(e,"1994-09-30":date);
bat.append(e,"1995-04-20":date);
bat.append(e,"1995-07-17":date);
bat.append(e,"1995-07-06":date);
bat.append(e,"1994-02-19":date);
bat.append(e,"1994-02-21":date);
bat.append(e,"1997-04-17":date);
bat.append(e,"1997-01-27":date);
bat.append(e,"1997-02-20":date);
bat.append(e,"1997-03-18":date);
bat.append(e,"1997-04-19":date);
bat.append(e,"1997-01-25":date);
bat.append(e,"1998-07-04":date);
bat.append(e,"1998-06-26":date);
bat.append(e,"1998-08-13":date);
bat.append(e,"1998-06-27":date);
bat.append(e,"1998-06-19":date);
bat.append(e,"1998-08-11":date);
bat.append(e,"1998-06-24":date);
bat.append(e,"1994-08-17":date);
bat.append(e,"1994-08-24":date);
bat.append(e,"1994-07-02":date);
bat.append(e,"1994-06-06":date);
bat.append(e,"1994-07-31":date);
bat.append(e,"1994-10-03":date);
bat.append(e,"1994-01-12":date);
bat.append(e,"1994-03-03":date);
bat.append(e,"1994-01-26":date);
bat.append(e,"1994-03-17":date);
bat.append(e,"1994-02-13":date);
bat.append(e,"1994-01-26":date);
bat.append(e,"1998-04-10":date);
bat.append(e,"1998-05-23":date);
bat.append(e,"1998-02-23":date);
bat.append(e,"1998-04-12":date);
bat.append(e,"1998-01-29":date);
bat.append(e,"1998-03-05":date);
bat.append(e,"1994-07-19":date);
bat.append(e,"1994-06-03":date);
bat.append(e,"1993-04-01":date);
bat.append(e,"1993-04-13":date);
bat.append(e,"1993-05-14":date);
bat.append(e,"1994-12-24":date);
bat.append(e,"1994-12-01":date);
bat.append(e,"1994-12-30":date);

xe:= mosaic.compress(e,"frame");
mosaic.dump(xe);
zd:= mosaic.decompress(xe);
io.print(e,zd);
#mosaic.analyse("frame prefix");
