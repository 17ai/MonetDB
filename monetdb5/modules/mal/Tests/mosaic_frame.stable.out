stdout of test 'mosaic_frame` in directory 'monetdb5/modules/mal` itself:


# 07:42:13 >  
# 07:42:13 >  "mserver5" "--debug=10" "--set" "gdk_nr_threads=0" "--set" "mapi_open=true" "--set" "mapi_port=30639" "--set" "mapi_usock=/var/tmp/mtest-1108/.s.monetdb.30639" "--set" "monet_prompt=" "--forcemito" "--set" "mal_listing=2" "--dbpath=/export/scratch1/mk/mosaic//Linux/var/MonetDB/mTests_monetdb5_modules_mal" "mosaic_frame.mal"
# 07:42:13 >  

# MonetDB 5 server v11.20.0
# This is an unreleased version
# Serving database 'mTests_monetdb5_modules_mal', using 8 threads
# Compiled for x86_64-unknown-linux-gnu/64bit with 64bit OIDs and 128bit integers dynamically linked
# Found 15.590 GiB available main-memory.
# Copyright (c) 1993-July 2008 CWI.
# Copyright (c) August 2008-2014 MonetDB B.V., all rights reserved
# Visit http://www.monetdb.org/ for further information
# Listening for connection requests on mapi:monetdb://vienna.ins.cwi.nl:30639/
# Listening for UNIX domain connection requests on mapi:monetdb:///var/tmp/mtest-1108/.s.monetdb.30639
# MonetDB/GIS module loaded
# MonetDB/SQL module loaded
function user.main():void;
    b := bat.new(:oid,:timestamp);
    bat.append(b,"2014-08-14 12:00:00.000":timestamp);
    bat.append(b,"2014-08-14 12:00:00.000":timestamp);
    bat.append(b,"2014-08-14 12:01:00.000":timestamp);
    bat.append(b,"2014-08-14 12:02:00.000":timestamp);
    bat.append(b,"2014-08-14 12:00:00.000":timestamp);
    b := bat.append(b,b);
    base := algebra.copy(b);
    io.print(base);
    x := mosaic.compress(b,"frame");
#mosaic.dump(x); 
    z := mosaic.decompress(x);
    io.print(z);
    d := bat.new(:oid,:date);
    bat.append(d,"2014-08-14":date);
    bat.append(d,"2014-08-14":date);
    bat.append(d,"2014-08-14":date);
    bat.append(d,"2014-08-14":date);
    bat.append(d,"2014-08-15":date);
    d := bat.append(d,d);
    io.print(d);
    xd := mosaic.compress(d,"frame");
#mosaic.dump(xd); 
    zd := mosaic.decompress(xd);
    io.print(zd);
end user.main;
#--------------------------#
# h	t  # name
# void	timestamp  # type
#--------------------------#
[ 0@0,	2014-08-14 12:00:00.000  ]
[ 1@0,	2014-08-14 12:00:00.000  ]
[ 2@0,	2014-08-14 12:01:00.000  ]
[ 3@0,	2014-08-14 12:02:00.000  ]
[ 4@0,	2014-08-14 12:00:00.000  ]
[ 5@0,	2014-08-14 12:00:00.000  ]
[ 6@0,	2014-08-14 12:00:00.000  ]
[ 7@0,	2014-08-14 12:01:00.000  ]
[ 8@0,	2014-08-14 12:02:00.000  ]
[ 9@0,	2014-08-14 12:00:00.000  ]
#--------------------------#
# h	t  # name
# void	timestamp  # type
#--------------------------#
[ 0@0,	2014-08-14 12:00:00.000  ]
[ 1@0,	2014-08-14 12:00:00.000  ]
[ 2@0,	2014-08-14 12:01:00.000  ]
[ 3@0,	2014-08-14 12:02:00.000  ]
[ 4@0,	2014-08-14 12:00:00.000  ]
[ 5@0,	2014-08-14 12:00:00.000  ]
[ 6@0,	2014-08-14 12:00:00.000  ]
[ 7@0,	2014-08-14 12:01:00.000  ]
[ 8@0,	2014-08-14 12:02:00.000  ]
[ 9@0,	2014-08-14 12:00:00.000  ]
#--------------------------#
# h	t  # name
# void	date  # type
#--------------------------#
[ 0@0,	2014-08-14  ]
[ 1@0,	2014-08-14  ]
[ 2@0,	2014-08-14  ]
[ 3@0,	2014-08-14  ]
[ 4@0,	2014-08-15  ]
[ 5@0,	2014-08-14  ]
[ 6@0,	2014-08-14  ]
[ 7@0,	2014-08-14  ]
[ 8@0,	2014-08-14  ]
[ 9@0,	2014-08-15  ]
#--------------------------#
# h	t  # name
# void	date  # type
#--------------------------#
[ 0@0,	2014-08-14  ]
[ 1@0,	2014-08-14  ]
[ 2@0,	2014-08-14  ]
[ 3@0,	2014-08-14  ]
[ 4@0,	2014-08-15  ]
[ 5@0,	2014-08-14  ]
[ 6@0,	2014-08-14  ]
[ 7@0,	2014-08-14  ]
[ 8@0,	2014-08-14  ]
[ 9@0,	2014-08-15  ]

# 07:42:13 >  
# 07:42:13 >  "Done."
# 07:42:13 >  

